File: app/src/main/java/com/fazemeright/chatbotmetcs622/workers/FireBaseSyncWorker.java
package com.fazemeright.chatbotmetcs622.workers;

import android.content.Context;
import androidx.annotation.NonNull;
import androidx.work.Worker;
import androidx.work.WorkerParameters;
import com.fazemeright.chatbotmetcs622.repositories.MessageRepository;

public class FireBaseSyncWorker extends Worker {
  public FireBaseSyncWorker(@NonNull Context context, @NonNull WorkerParameters workerParams) {
    super(context, workerParams);
  }

  @NonNull
  @Override
  public Result doWork() {
    syncMessagesWithCloudAndLocal();
    return Result.success();
  }

  /**
   * Sync messages from local to cloud, and then from cloud to local
   */
  private void syncMessagesWithCloudAndLocal() {
    final MessageRepository messageRepository =
        MessageRepository.getInstance(getApplicationContext());
    messageRepository.addMessagesToFireBase(messageRepository.getAllMessages());

    messageRepository.syncMessagesFromFireStoreToRoom();
  }
}

File: app/src/main/java/com/fazemeright/chatbotmetcs622/workers/FireBaseSyncWorker.kt
package com.fazemeright.chatbotmetcs622.workers

import android.app.Application
import android.content.Context
import androidx.work.CoroutineWorker
import androidx.work.WorkerParameters
import com.fazemeright.chatbotmetcs622.domain.SyncLocalAndCloudMessagesUseCase

class FireBaseSyncWorker(context: Context, workerParams: WorkerParameters) : CoroutineWorker(context, workerParams) {
    private val syncLocalAndCloudMessagesUseCase by lazy { SyncLocalAndCloudMessagesUseCase(context as Application) }

    override suspend fun doWork(): Result {
        syncLocalAndCloudMessagesUseCase()
        return Result.success()
    }
}